Src/GoogleApis.Tests/Apis/Requests/PageStreamerTest.cs

```diff
@@ -0,0 +1,241 @@
+                new PageStreamer<int, Request, Page, string>(
+                    (request, token) => request.WithToken(token),
+                    page => page.NextPageToken,
+                    page => page.Items);
+
+            // Task to await when fetching pages; defaults to an already completed task,
+            // but can be set to another task to allow for more fine-grained interaction.
+            internal Task gateKeeper = TaskEx.FromResult(0);
+            internal List<Page> Pages { get; private set; }
+            internal string Name { get; set; }
+            internal string RequestCheck { get; set; }
+
+            internal PageStreamedResource(string name, params Page[] pages)
+            {                
+                this.Name = name;
+                this.Pages = pages.ToList();
+            }
+
+            public override string ToString()
+            {
+                return Name;
+            }
+
+            internal Request NewRequest()
+            {
+                return new Request(this) { Check = RequestCheck };
+            }
+
+            internal Page GetPage(string token, string check)
+            {
+                Assert.AreEqual(RequestCheck, check);
+                int index = Pages.FindIndex(page => page.NextPageToken == token);
+                // A request for the first page will have the NextPageToken of the last page... so we should always
+                // find something.
+                Assert.AreNotEqual(-1, index);
+                index = (index + 1) % Pages.Count;
+                return Pages[index];
+            }
+
+            internal async Task<Page> GetPageAsync(string token, string check, CancellationToken cancellationToken)
+            {
+                await gateKeeper;
+                cancellationToken.ThrowIfCancellationRequested();
+                return GetPage(token, check);
+            }
+
+            public IEnumerable<int> AllItems => Pages.SelectMany(page => page.Items);
+        }
+
+        /// <summary>
+        /// Request to work with <see cref="PageStreamedResource"/>; feels like overkill (in a non-HTTP context)
+        /// due to having to implement <see cref="IClientServiceRequest{Page}"/> but there isn't an interface
+        /// which only includes <c>Execute</c> and <c>ExecuteAsync.</c>
+        /// </summary>
+        internal class Request : IClientServiceRequest<Page>
+        {
+            internal string Token { get; set; }
+            // Used to check that we keep the right request info
+            internal string Check { get; set; }
+
+            private readonly PageStreamedResource resource;
+
+            internal Request(PageStreamedResource resource)
+            {
+                this.resource = resource;
+            }
+
+            internal Request WithToken(string token)
+            {
+                return new Request(resource) { Token = token, Check = Check };
```

Comment:
Do we need to create a new Request here? Why not just updating the Token and Check fields?
In real usage, we are just going to set the NextPageToken on the current request, so I would recommend doing something similar here.